ks_patches/vaksine_utils.js

```diff
@@ -0,0 +1,170 @@
+    if(doseNr > 3 || doseNr < 1) {
+        return undefined;
+    }
+    var ids = getVaksineIds(doseNr);
+    return getNameAndDate(ids, selectedTei.attributes);
+}
+
+function getVaksineIds(doseNr) {
+    if(doseNr === 1) {
+        return {
+            typeId: PROFIL_VAKSINE_1_TYPE_ID,
+            dateId: PROFIL_VAKSINE_1_DATO_ID
+        }
+    }
+    if(doseNr === 2) {
+        return {
+            typeId: PROFIL_VAKSINE_2_TYPE_ID,
+            dateId: PROFIL_VAKSINE_2_DATO_ID
+        }
+    }
+    if(doseNr === 3) {
+        return {
+            typeId: PROFIL_VAKSINE_3_TYPE_ID,
+            dateId: PROFIL_VAKSINE_3_DATO_ID
+        }
+    }
+}
+
+function getNameAndDate(ids, attributes) {
+    var dose = {
+        name: attributes.find(att => att.attribute === ids.typeId).value,
+        date: attributes.find(att => att.attribute === ids.dateId).value
+    };
+    return dose && dose.date || dose.name ? dose : undefined;
+}
+
+function convertVaccineDate(vaccinationDate, DateUtils) {
+    const tmpDate = angular.copy(vaccinationDate);
+    if(tmpDate[1]) {
+        tmpDate[1] = tmpDate[1] - 1;
+    }
+    return DateUtils.getDateFromUTCString(tmpDate);
+}
+
+function getVaccineNicenameAndType(vaccine) {
+    switch(vaccine.vaccineCode.code) {
+        case "ASZ03":
+            return {
+                name: "Vaxzevria (AstraZeneca)",
+                type: "AstraZeneca"
+            };
+        case "MOD03":
+            return {
+                name: "Covid-19 Vaccine Moderna",
+                type: "Moderna"
+            };
+        case "BNT03":
+            return {
+                name: "Comirnaty (BioNTech og Pfizer)",
+                type: "Comirnaty"
+            };
+        case "JAN03":
+            return {
+                name: "Covid-19 Vaccine Janssen",
+                type: "Janssen"
+            };
+        case "CUR03":  // Not supported in DHIS2 yet
+        default:
+            return {
+            name: vaccine.vaccineCode.display + ' (Annet)',
+                type: "annenvaksine"
+            };
+    }
+}
+
+
+function enrichWithValidation(vaccine) {
+    if(!vaccine.profileDose) {
+        return {...vaccine, vaccineIsInProfile: false, dateMismatch: false, nameMismatch: false, updatePossible: true};
+    }
+    var dateMismatch = vaccine.date !== vaccine.profileDose.date;
+    var nameMismatch = vaccine.name !== vaccine.profileDose.name && vaccine.type !== vaccine.profileDose.name;
+    return {...vaccine,
+        vaccineIsInProfile: true,
+        dateMismatch,
+        nameMismatch,
+        updatePossible:  dateMismatch || nameMismatch
+    }
+}
+
+export function saveVaccineToProfile(tei, vaccines, attributesById, TEIService, $q) {
+
+    var teiCopy = angular.copy(tei);
+
+    teiCopy.attributes = getUpdatedVaccineAttributes(teiCopy.attributes, vaccines);
+    if(!tei.trackedEntityInstance) { // If trackedEntityInstance is not set, we assume we are registering
```

Comment:
Can this code be called from the registration form? In that case, will not the TEIService.update below be a problem?

Reply:
This _is_ the suboptimal solution to that problem after Helge relized the problem. 
TEIService.update will now NOT be called if you are on the registration form (i.e. tei.trackedEntityInstance is nullish). Then we just update the form without sending a call to the backend.

Reply:
Ah yes now I see the return statement preventing the TEIService to be called.

Reply:
Could it be better to split this into two functions and call the "right" one from the form instead?

Reply:
The form does not know the context, as it is called from the modal. But I could try to push the logic a bit more forward and split it up to make it more clear!